File: packages/target-dotnet/src/Naming.cs
using System.Globalization;
using System.Text;

namespace Besdk.Target.Dotnet;

/// <summary>
/// Turning IR token sequences into C# identifiers.
/// </summary>
/// <remarks>
/// <para>
/// .NET is the only target here that capitalises <i>members</i>: methods, properties, and events are all
/// <c>PascalCase</c>, where every other language uses <c>camelCase</c> for at least some of them. Parameters and
/// locals stay <c>camelCase</c>.
/// </para>
/// <para>
/// That makes this the clearest payoff of names-as-token-sequences: the same IR name becomes <c>UserId</c> here,
/// <c>UserID</c> in Go, <c>userId</c> in Java and TypeScript, and <c>user_id</c> in Python — with no coordination
/// between the targets.
/// </para>
/// </remarks>
internal static class Naming
{
  /// <summary>
  /// C#'s reserved words.
  /// </summary>
  /// <remarks>
  /// Contextual keywords — <c>record</c>, <c>async</c>, <c>value</c>, <c>required</c> — are deliberately absent:
  /// they are legal identifiers, and renaming a property the spec called <c>value</c> would be gratuitous.
  /// </remarks>
  private static readonly HashSet<string> Reserved = new(StringComparer.Ordinal)
    {
        "abstract", "as", "base", "bool", "break", "byte", "case", "catch", "char", "checked", "class", "const",
        "continue", "decimal", "default", "delegate", "do", "double", "else", "enum", "event", "explicit",
        "extern", "false", "finally", "fixed", "float", "for", "foreach", "goto", "if", "implicit", "in", "int",
        "interface", "internal", "is", "lock", "long", "namespace", "new", "null", "object", "operator", "out",
        "override", "params", "private", "protected", "public", "readonly", "ref", "return", "sbyte", "sealed",
        "short", "sizeof", "stackalloc", "static", "string", "struct", "switch", "this", "throw", "true", "try",
        "typeof", "uint", "ulong", "unchecked", "unsafe", "ushort", "using", "virtual", "void", "volatile",
        "while",
    };

  /// <summary>A type or member name: <c>PascalCase</c>.</summary>
  public static string Pascal(IReadOnlyList<string> tokens)
  {
    var builder = new StringBuilder();
    foreach (var token in tokens)
    {
      builder.Append(Capitalise(token));
    }

    var name = builder.ToString();
    if (name.Length == 0)
    {
      return "Value";
    }

    if (char.IsDigit(name[0]))
    {
      // `2FactorAuth` is not an identifier. Prefixed rather than stripped, because the digit is usually
      // meaningful.
      return "N" + name;
    }

    // A PascalCase name can only collide with a keyword if the keyword is capitalised, which none are — but a
    // single-token name like `class` becomes `Class`, which is fine. Kept as a guard for the tokenless case.
    return name;
  }

  /// <summary>A parameter or local: <c>camelCase</c>, escaped when it collides with a keyword.</summary>
  public static string Camel(IReadOnlyList<string> tokens)
  {
    var pascal = Pascal(tokens);
    var name = char.ToLowerInvariant(pascal[0]) + pascal[1..];
    if (char.IsDigit(name[0]))
    {
      return "n" + pascal;
    }

    // `@` rather than a rename: C# has a verbatim-identifier escape precisely for this, and `@class` reads
    // better than `classValue` while keeping the spec's own word.
    return Reserved.Contains(name) ? "@" + name : name;
  }

  /// <summary>
  /// A namespace from a package name.
  /// </summary>
  /// <remarks>
  /// <c>Acme.Sdk</c> from <c>Acme.Sdk</c>, and <c>Acme.WidgetSdk</c> from <c>acme-widget-sdk</c>. Each segment is
  /// PascalCased, because .NET namespaces are PascalCase by convention — unlike Java's all-lowercase.
  /// </remarks>
  public static string Namespace(string value)
  {
    var parts = new List<string>();
    foreach (var segment in value.Split('.', '/', ':'))
    {
      var words = segment.Split('-', '_', ' ');
      var joined = new StringBuilder();
      foreach (var word in words)
      {
        joined.Append(Capitalise(word));
      }

      if (joined.Length > 0)
      {
        parts.Add(char.IsDigit(joined[0]) ? "N" + joined : joined.ToString());
      }
    }

    return parts.Count == 0 ? "Sdk" : string.Join('.', parts);
  }

  /// <summary>An enum member: <c>PascalCase</c>, which is .NET's convention rather than SCREAMING_SNAKE.</summary>
  public static string EnumMember(IReadOnlyList<string> tokens) => Pascal(tokens);

  private static string Capitalise(string token)
  {
    var clean = new string(token.Where(char.IsLetterOrDigit).ToArray());
    if (clean.Length == 0)
    {
      return string.Empty;
    }

    return char.ToUpperInvariant(clean[0]) + clean[1..].ToLowerInvariant() is var lowered && clean.Length > 1
        // Preserve interior capitals when the token already has them (`OpenAI` stays `OpenAI`), but
        // lowercase an all-caps token so `API` becomes `Api` — which is .NET's own convention for
        // initialisms longer than two letters.
        ? char.ToUpperInvariant(clean[0]) + (clean[1..].All(char.IsUpper) ? clean[1..].ToLowerInvariant() : clean[1..])
        : lowered;
  }

  /// <summary>Uppercase for an environment variable: <c>ACME_TOKEN</c>.</summary>
  public static string ScreamingSnake(IReadOnlyList<string> tokens) =>
      string.Join('_', tokens.Select(token => token.ToUpperInvariant()));
}

File: packages/target-php/src/Naming.php
<?php

declare(strict_types=1);

namespace Graft\Target\Php;

/**
 * Turning IR token sequences into PHP identifiers.
 *
 * Names arrive as lowercase token sequences (`["user","id"]`) precisely so each target applies its own
 * convention (SPEC.md §3.2). PHP's is PSR-1 and PSR-12: `PascalCase` for classes, `camelCase` for methods
 * and properties.
 *
 * Unlike Go, PHP does not capitalise initialisms wholly — `getApiKey`, not `getAPIKey` — which is one of
 * the places the token-sequence design pays off: the same IR produces `UserID` in Go and `userId` here
 * with no coordination.
 */
final class Naming
{
    /**
     * Words PHP reserves. A generated identifier colliding with one of these is a syntax error, not a
     * style problem, so the list is exhaustive for 8.4 rather than a best guess.
     *
     * @var list<string>
     */
    private const RESERVED = [
        'abstract', 'and', 'array', 'as', 'break', 'callable', 'case', 'catch', 'class', 'clone',
        'const', 'continue', 'declare', 'default', 'do', 'echo', 'else', 'elseif', 'empty',
        'enddeclare', 'endfor', 'endforeach', 'endif', 'endswitch', 'endwhile', 'enum', 'eval',
        'exit', 'extends', 'final', 'finally', 'fn', 'for', 'foreach', 'function', 'global', 'goto',
        'if', 'implements', 'include', 'include_once', 'instanceof', 'insteadof', 'interface', 'isset',
        'list', 'match', 'namespace', 'new', 'or', 'print', 'private', 'protected', 'public',
        'readonly', 'require', 'require_once', 'return', 'static', 'switch', 'throw', 'trait', 'try',
        'unset', 'use', 'var', 'while', 'xor', 'yield',
        // Not reserved as identifiers but reserved as *type* names, which a class cannot be called.
        'bool', 'float', 'int', 'iterable', 'mixed', 'never', 'null', 'object', 'string', 'void',
        'false', 'true', 'self', 'parent',
    ];

    /**
     * @param list<string> $tokens
     */
    public static function pascal(array $tokens): string
    {
        $out = '';
        foreach ($tokens as $token) {
            $out .= self::capitalise($token);
        }

        return self::safeClass($out);
    }

    /**
     * @param list<string> $tokens
     */
    public static function camel(array $tokens): string
    {
        $pascal = '';
        foreach ($tokens as $token) {
            $pascal .= self::capitalise($token);
        }

        return self::safeMember(lcfirst($pascal));
    }

    /**
     * A property name, which follows the wire name only when it has to.
     *
     * A wire key that is not a valid PHP identifier — `_id`, `2fa`, `content-type` — still has to be
     * readable, so the property is named idiomatically and the mapping to the wire happens in the decoder.
     * Never renaming would produce `$widget->{'content-type'}`, which is legal and awful.
     *
     * @param list<string> $tokens
     */
    public static function property(array $tokens): string
    {
        return self::camel($tokens);
    }

    /** Class-name-safe: a leading digit or a reserved word gets a prefix rather than a mangled name. */
    private static function safeClass(string $name): string
    {
        if ($name === '') {
            return 'Value';
        }
        if (preg_match('/^\d/', $name) === 1) {
            // `2FactorAuth` is not a class name. Prefixed rather than stripped, because the digit is
            // usually meaningful.
            return 'N' . $name;
        }
        if (in_array(strtolower($name), self::RESERVED, true)) {
            return $name . 'Type';
        }

        return $name;
    }

    private static function safeMember(string $name): string
    {
        if ($name === '') {
            return 'value';
        }
        if (preg_match('/^\d/', $name) === 1) {
            return 'n' . ucfirst($name);
        }
        // Method and property names may be reserved words in PHP — `$obj->list` and `$obj->class()` are
        // both legal — so only the ones that are genuinely ambiguous are renamed. Kept minimal: renaming
        // a member the spec named `list` would be gratuitous.
        return $name;
    }

    private static function capitalise(string $token): string
    {
        // A token may itself contain characters PHP forbids, e.g. a wire key like `content-type` that was
        // never tokenised. Stripped rather than escaped: an identifier has to be an identifier.
        $clean = preg_replace('/[^A-Za-z0-9]/', '', $token) ?? '';

        return ucfirst($clean);
    }

    /**
     * A PSR-4 namespace from a package name.
     *
     * `acme/widget-sdk` becomes `Acme\WidgetSdk`, which is the mapping Composer's own convention implies
     * and what every PHP consumer will expect from that package name.
     */
    public static function namespaceFor(string $packageName): string
    {
        $parts = [];
        foreach (explode('/', $packageName) as $segment) {
            $words = preg_split('/[^A-Za-z0-9]+/', $segment) ?: [];
            $joined = '';
            foreach ($words as $word) {
                $joined .= ucfirst(strtolower($word));
            }
            if ($joined !== '') {
                $parts[] = $joined;
            }
        }

        return $parts === [] ? 'Sdk' : implode('\\', $parts);
    }
}
